include/gdf/cffi/types.h

```diff
@@ -20,86 +24,136 @@ typedef enum {
+ *
+ * @note held either at the single-element or whole-column level.
+ *
+ */
 typedef struct {
 	gdf_time_unit time_unit;
 	// here we can also hold info for decimal datatype or any other datatype that requires additional information
 } gdf_dtype_extra_info;
 
-typedef struct gdf_column_{
-    void *data;                       /**< Pointer to the columns data */ 
-    gdf_valid_type *valid;            /**< Pointer to the columns validity bit mask where the 'i'th bit indicates if the 'i'th row is NULL */
-    gdf_size_type size;               /**< Number of data elements in the columns data buffer*/
-    gdf_dtype dtype;                  /**< The datatype of the column's data */
-    gdf_size_type null_count;         /**< The number of NULL values in the column's data */
-    gdf_dtype_extra_info dtype_info;
-    char *			col_name;			// host-side:	null terminated string
-} gdf_column;
-
-/* --------------------------------------------------------------------------*/
+/**
+ * @brief The fundamental, columnar format of data the GDF library works with.
+ *
+ * A gdf_column_ may originate in a RDBMS-like schema table; it may be the intermediary result
+ * within an execution plan; or it may be the result of non-DBMS related computation.
+ */
+ typedef struct gdf_column_{
+     void *data;
+         ///< Type-erased pointer to the column's raw data - which is a consecutive sequence
+         ///< with no gaps of elements of the type represented by `dtype`.
+         ///<
+         ///< @todo There are currently no formal alignment requirements, but it seems the
+         ///< implementation may implicitly be assuming alignment to the size of the relevant
+         ///< type.
+         ///< @todo Can this be NULL? What about after "construction"?
+         ///< @todo Is this always in device memory?
+
+     gdf_valid_type *valid;
+         ///< a pseudo-column of `size` bits, packed into bytes (with in-byte order from
+         ///< the least significant to the most significant bit), indicating whether
+         ///< the column element is null (bit value is 0) or not null (bit value is 1;
+         ///< a "valid" element)
+         ///<
+         ///< @todo There are currently no formal alignment requirements, but it seems the
+         ///< implementation may implicitly be assuming alignment to the size of the relevant
+         ///<
+         ///< @todo Is this expressly forbidden from being NULL in the case of a 0 null-count?
+
+     gdf_size_type size;
+         ///< The number of column elements (_not_ their total size in bytes, _nor_ the
+         ///< size of an individual element)
+         ///<
+         ///< @todo is it allocated capacity or size in use?
+
+     gdf_dtype dtype;
+         ///< An indicator of the column's data type, for type un-erasure
+
+     gdf_size_type null_count;
+         ///< The number of null elements in the column, which is
+         ///< also the number of 1 bits in the `valid` pseudo-column
```

Comment:
`null_count` would be the number of `0` bits in the range `[0, size)`.

Reply:
@jrhemstad : Right, fixed.